src/cipher.rs

```diff
@@ -472,319 +486,651 @@ impl rustls_root_cert_store {
+/// e.g. `rustls_web_pki_client_cert_verifier_builder_add_crl`. Once you're
+/// done configuring settings, call `rustls_web_pki_client_cert_verifier_builder_build`
+/// to turn it into a `rustls_client_cert_verifier`. This object is not safe
+/// for concurrent mutation.
+// TODO(@cpu): Add rustdoc link once available.
+pub struct rustls_web_pki_client_cert_verifier_builder {
+    // We use the opaque struct pattern to tell C about our types without
+    // telling them what's inside.
+    // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
     _private: [u8; 0],
 }
 
-impl Castable for rustls_allow_any_authenticated_client_builder {
+pub(crate) struct ClientCertVerifierBuilder {
+    roots: Arc<RootCertStore>,
+    root_hint_subjects: Vec<DistinguishedName>,
+    crls: Vec<CertificateRevocationListDer<'static>>,
+    revocation_depth: RevocationCheckDepth,
+    revocation_policy: UnknownStatusPolicy,
+    allow_anonymous: bool,
+}
+
+impl Castable for rustls_web_pki_client_cert_verifier_builder {
     type Ownership = OwnershipBox;
-    // NOTE: contained value is consumed even on error, so this can contain None, but the caller
-    // still needs to free it
-    type RustType = Option<AllowAnyAuthenticatedClient>;
+    type RustType = Option<ClientCertVerifierBuilder>;
 }
 
-impl rustls_allow_any_authenticated_client_builder {
-    /// Create a new allow any authenticated client certificate verifier builder using the root store.
+impl rustls_web_pki_client_cert_verifier_builder {
+    /// Create a `rustls_web_pki_client_cert_verifier_builder`. Caller owns the memory and must
+    /// eventually call `rustls_web_pki_client_cert_verifier_builder_build`, then free the
+    /// resulting `rustls_client_cert_verifier`.
+    ///
+    /// Without further modification the builder will produce a client certificate verifier that
+    /// will require a client present a client certificate that chains to one of the trust anchors
+    /// in the provided `rustls_root_cert_store`. The root cert store must not be empty.
     ///
-    /// This copies the contents of the rustls_root_cert_store. It does not take
-    /// ownership of the pointed-to memory.
+    /// Revocation checking will not be performed unless
+    /// `rustls_web_pki_client_cert_verifier_builder_add_crl` is used to add certificate revocation
+    /// lists (CRLs) to the builder. If CRLs are added, revocation checking will be performed
+    /// for the entire certificate chain unless
+    /// `rustls_web_pki_client_cert_verifier_only_check_end_entity_revocation` is used. Unknown
+    /// revocation status for certificates considered for revocation status will be treated as
+    /// an error unless `rustls_web_pki_client_cert_verifier_allow_unknown_revocation_status` is
+    /// used.
     ///
-    /// This object can then be used to load any CRLs.
+    /// Anonymous unauthenticated clients will not be permitted unless
```

Comment:
```suggestion
    /// Unauthenticated clients will not be permitted unless
```

I think the change from `anonymous` to `unauthenticated` is a good one; this is a missed spot.

Reply:
Fixed up throughout.